misc-utils/fincore.c

```diff
@@ -42,40 +43,93 @@
 #define N_PAGES_IN_WINDOW ((size_t)(32 * 1024))
 
+#ifndef HAVE_CACHESTAT
+
+#include <linux/types.h>
+
+#ifndef __NR_cachestat
+#define __NR_cachestat 451
+#endif
+
+struct cachestat_range {
+	__u64 off;
+	__u64 len;
+};
+
+struct cachestat {
+	__u64 nr_cache;
+	__u64 nr_dirty;
+	__u64 nr_writeback;
+	__u64 nr_evicted;
+	__u64 nr_recently_evicted;
+};
+
+static inline int cachestat(unsigned int fd,
+			    const struct cachestat_range *cstat_range,
+			    struct cachestat *cstat, unsigned int flags)
+{
+	return syscall(__NR_cachestat, fd, cstat_range, cstat, flags);
+}
+
+#endif // HAVE_CACHESTAT
 
 struct colinfo {
 	const char * const name;
 	double whint;
 	int flags;
 	const char *help;
+	unsigned int pages : 1;
 };
 
 enum {
 	COL_PAGES,
 	COL_SIZE,
 	COL_FILE,
-	COL_RES
+	COL_RES,
+	COL_CACHE_PAGES,
+	COL_CACHE,
+	COL_DIRTY_PAGES,
+	COL_DIRTY,
+	COL_WRITEBACK_PAGES,
+	COL_WRITEBACK,
+	COL_EVICTED_PAGES,
+	COL_EVICTED,
+	COL_RECENTLY_EVICTED_PAGES,
+	COL_RECENTLY_EVICTED,
 };
 
 static const struct colinfo infos[] = {
-	[COL_PAGES]  = { "PAGES",    1, SCOLS_FL_RIGHT, N_("file data resident in memory in pages")},
-	[COL_RES]    = { "RES",      5, SCOLS_FL_RIGHT, N_("file data resident in memory in bytes")},
-	[COL_SIZE]   = { "SIZE",     5, SCOLS_FL_RIGHT, N_("size of the file")},
-	[COL_FILE]   = { "FILE",     4, 0, N_("file name")},
+	[COL_PAGES]                  = { "PAGES",                  1, SCOLS_FL_RIGHT, N_("file data resident in memory in pages"), 1},
+	[COL_RES]                    = { "RES",                    5, SCOLS_FL_RIGHT, N_("file data resident in memory in bytes")},
+	[COL_SIZE]                   = { "SIZE",                   5, SCOLS_FL_RIGHT, N_("size of the file")},
+	[COL_FILE]                   = { "FILE",                   4, 0, N_("file name")},
+	[COL_CACHE_PAGES]            = { "CACHE_PAGES",            1, SCOLS_FL_RIGHT, N_("number of cached pages"), 1},
+	[COL_CACHE]                  = { "CACHE",                  5, SCOLS_FL_RIGHT, N_("number of cached bytes")},
+	[COL_DIRTY_PAGES]            = { "DIRTY_PAGES",            1, SCOLS_FL_RIGHT, N_("number of dirty pages"), 1},
+	[COL_DIRTY]                  = { "DIRTY",                  5, SCOLS_FL_RIGHT, N_("number of dirty bytes")},
+	[COL_WRITEBACK_PAGES]        = { "WRITEBACK_PAGES",        1, SCOLS_FL_RIGHT, N_("number of pages marked for writeback"), 1},
+	[COL_WRITEBACK]              = { "WRITEBACK",              5, SCOLS_FL_RIGHT, N_("number of bytes marked for writeback")},
+	[COL_EVICTED_PAGES]          = { "EVICTED_PAGES",          1, SCOLS_FL_RIGHT, N_("number of evicted pages"), 1},
+	[COL_EVICTED]                = { "EVICTED",                5, SCOLS_FL_RIGHT, N_("number of evicted bytes")},
+	[COL_RECENTLY_EVICTED_PAGES] = { "RECENTLY_EVICTED_PAGES", 1, SCOLS_FL_RIGHT, N_("number of recently evicted pages"), 1},
+	[COL_RECENTLY_EVICTED]       = { "RECENTLY_EVICTED",       5, SCOLS_FL_RIGHT, N_("number of recently evicted bytes")},
```

Comment:
The names are soooo long :-)

Reply:
I'm bad at naming things :(

All abbreviations I can come up with and sound bad.
I always hope for you to complain and come up with something better in the process.

... waiting patiently.

Reply:
We could also drop all columns of a certain type.
Then users can calculate it on their own if they really need it.

A utility in util-linux to print the current pagesize seems useful, too.
Maybe just `fincore --pagesize`.